src/main/scala/devsearch/features/QueryRecognizer.scala

```diff
@@ -0,0 +1,37 @@
+package devsearch.features
+
+import devsearch.parsers.Languages
+
+object QueryRecognizer {
+  def apply(query: String): Option[CodeFile] = {
+    scala.util.Try {
+      Languages.supportedLanguages().toList.map { language =>
+        // Some parsers (Go) need a valid file name to parse correctly, so we
+        // generate one for the query.
+        val generatedFileName = "file" + (Languages.extension(language) match {
+          case Some(extension) => "." + extension
+          case None => ""
+        })
+
+        CodeFile(
+          language,
+          CodeFileLocation("dummy_user", "dummy_repo", generatedFileName),
+          query
+        )
+      }.map { codeFile =>
+        // Scala parser is more permissive, so we reduce its score for the comparison.
+        val weight: Double = codeFile.language match {
+          case Languages.Scala => 0.7
+          case _ => 1.0
+        }
+        val score: Double = FeatureRecognizer(codeFile).size.toDouble * weight
+        (codeFile, score)
+      }.sortBy(_._2)
+      .reverse.head match {
+        case (codeFile, score) if score > 0.0 => Some(codeFile)
+        case _ => None
+      }
+      // head can throw NoSuchElementException
```

Comment:
Since supported language is non-empty and you call no filter, head is always defined. Also, in general, it is better to use headOption than wrapping in a Try since the throw-catch mechanism is quite expensive on the JVM.